PR: Create a simple code editor and use that for other plugins

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

This is a preliminary step in the migration of the Editor Plugin to the new API. 

The CodeEditor is a very complex widget and has been used by many other plugins for providing a code editor, where the only basic need seems to have syntax highlighting capabilities at most.

Currently, the following widgets use the CodeEditor

### Variable Explorer

Object Editor (For python highlight)

### Before

<img width="832" alt="Screen Shot 2020-08-03 at 8 30 17" src="https://user-images.githubusercontent.com/3627835/89187939-94042680-d563-11ea-8a63-bcc2ee2bfa2d.png">

### Now
<img width="832" alt="Screen Shot 2020-08-03 at 8 27 26" src="https://user-images.githubusercontent.com/3627835/89187653-2f48cc00-d563-11ea-833b-6a03544ab482.png">

## History Plugin
For python highlight

### Before

<img width="718" alt="Screen Shot 2020-08-03 at 8 30 43" src="https://user-images.githubusercontent.com/3627835/89187986-a2524280-d563-11ea-801a-7b4c9a87a548.png">

### Now
<img width="759" alt="Screen Shot 2020-08-03 at 8 25 46" src="https://user-images.githubusercontent.com/3627835/89187460-f27cd500-d562-11ea-8f29-944299a8fe72.png">

## Help Plugin

For the plain text view (For python highlight)

### Before

<img width="720" alt="Screen Shot 2020-08-03 at 8 31 10" src="https://user-images.githubusercontent.com/3627835/89188017-b26a2200-d563-11ea-9af8-4c7eca8e0cba.png">

### Now

<img width="756" alt="Screen Shot 2020-08-03 at 8 26 06" src="https://user-images.githubusercontent.com/3627835/89187499-ff99c400-d562-11ea-88cd-db7bc03b1053.png">

## Appearance plugin

For the Syntax Highlighter preview (For python highlight)

### Before

<img width="1162" alt="Screen Shot 2020-08-03 at 8 31 24" src="https://user-images.githubusercontent.com/3627835/89188053-bb5af380-d563-11ea-8347-84d5bad37676.png">

### Now

<img width="1162" alt="Screen Shot 2020-08-03 at 8 27 09" src="https://user-images.githubusercontent.com/3627835/89187606-235d0a00-d563-11ea-8e5b-5a48ddbbc36f.png">

## Error report

For markdown highlight. This one uses more of the editor cursor and selection capabilities, but only a handful.

### Before

<img width="682" alt="Screen Shot 2020-08-03 at 8 31 37" src="https://user-images.githubusercontent.com/3627835/89188076-c31a9800-d563-11ea-964c-a11a6333f6d5.png">

### Now

<img width="682" alt="Screen Shot 2020-08-03 at 8 26 25" src="https://user-images.githubusercontent.com/3627835/89187527-0a545900-d563-11ea-8b8b-6c15e5321c12.png">


As an initial attempt to really decouple the different plugins the simplest way (by far) is to provide a simple code editor with the needed capabilities in the `spyder.widgets` module. This boils down to syntax highlighting and a few options. No Panels, no Extensions etc...

This has the following benefits.
- **Performance**: Spyder now can live without 5 extra instances (in the worst case) of a complex (and heavier) widget like the fully blown code editor. This could be more if the History plugin eventually tracks consoles separately.
- **Plugin dependencies**: hardcoded required dependencies are now optional so Spyder can start without the Code Editor and have all the other plugins work normally (History, VarExp, Help, Appearance).
- **Cleaner code** that avoids circular (or impossible) dependencies between plugins. (The appearance plugin needs to be loaded before the editor, but it uses the editor...)


- The new API dependencies (except for the Editor) can now really be tested.
- With the new work on syntax highlighting, having a simple code editor to test things might help in the process.

### (minor) Drawbacks:

   * Appearance: some things are not visible now on this panel. But, this panel is just a preview and the most important thing is the actual colors, not having line numbers in this context seems like a really minor issue.

**Update**

The simple code editor now implements line numbers, so there is no longer any limitation with replacing CodeEditor by this SimpleCodeEditor on the before mentioned places.

Some small cleanup is needed still but so far

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`spyder/plugins/history/widgets.py`](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/plugins/history/widgets.py):

> [Line 276:1](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/plugins/history/widgets.py#L276): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace



* In the file [`spyder/widgets/editor.py`](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/widgets/editor.py):

> [Line 83:52](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/widgets/editor.py#L83): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 84:68](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/widgets/editor.py#L84): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 97:9](https://github.com/spyder-ide/spyder/blob/4adce00db1d20ae03440dc84ab2c0bab8064e090/spyder/widgets/editor.py#L97): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line

ccordoba12 commented:
> As an initial attempt to really decouple the different plugins the simplest way (by far) is to provide a simple code editor with the needed capabilities in the spyder.widgets module

Ok, I agree with this.

> Since this simple editor does not implement panels, no line numbers are available.

Can't you add line numbers to this simplified version of CodeEditor? We had a line number area in the old days of Spyder 2 without a panel, so that should be possible in this case too.

goanpeca commented:
>Can't you add line numbers to this simplified version of CodeEditor? We had a line number area in the old days of Spyder 2 without a panel, so that should be possible in this case too.

Yes it is possible. I will add one.

ccordoba12 commented:
Thanks @goanpeca!

goanpeca commented:
This is ready for review. Failure seems unrelated to the work of this PR.

ccordoba12 commented:
@goanpeca, please rebase on top of the latest master to fix the error with our tests.

ccordoba12 commented on spyder/plugins/appearance/confpage.py:
```diff
@@ -321,27 +319,32 @@ def update_preview(self, index=None, scheme_name=None):
         """
         text = ('"""A string"""\n\n'
                 '# A comment\n\n'
-                '# %% A cell\n\n'
```
Why are we not displaying cells in the preview anymore? That's a very important functionality, so I don't agree with removing them.

goanpeca replied:
It is not an important functionality for a simple code editor. Having cells appear in the appearance settings is completely secondary. I do not agree with implementing that.

ccordoba12 commented on spyder/plugins/appearance/confpage.py:
```diff
@@ -321,27 +319,32 @@ def update_preview(self, index=None, scheme_name=None):
-                                         scroll_past_end=update_scrollbar)
-        self.preview_editor.set_text(text)
+
+        # FIXME: The appearance plugin is not tied to the code editor
+        # so these calls here probably should not be used, since it
+        # assumes there is an editor, which might not be the case at
+        # all. We can probably make a decission for the user and
+        # that should be good enough, after all this is just a quick
+        # preview.
+        show_blanks = CONF.get('editor', 'blank_spaces')
+        update_scrollbar = CONF.get('editor', 'scroll_past_end')
```
Let's take those decisions now: please make `show_blanks` and `update_scrollbar` `False` in the preview.

goanpeca replied:
👍

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/plugins/help/widgets.py:
```diff
@@ -349,6 +353,18 @@ def setup(self, options):
             initial=self.get_option('wrap'),
         )
+        self.copy_action = self.create_action(
+            name="help_plain_text_copy_action",
```
Why is this action not listed in an enum class, like in other plugins moved to the new API?

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/plugins/help/widgets.py:
```diff
@@ -349,6 +353,18 @@ def setup(self, options):
+            register_shortcut=False,
+        )
+        self.select_all_action = self.create_action(
+            name="help_plain_text_select_all_action",
```
And this one?

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -420,20 +420,28 @@ def _update_details_for_item(self, tree_item):
             data = attr_details.data_fn(tree_item)
             self.editor.setPlainText(data)
             self.editor.setWordWrapMode(attr_details.line_wrap)
+            # FIXME: The variable explorer is not tied to the code editor
+            # so these calls here probably should not be used, since it
+            # assumes there is an editor, which might not be the case at
+            # all. We can probably make a decission for the user and
+            # that should be good. Having all code editors in spyder in
+            # sync does not seem very useful in this case.
             show_blanks = CONF.get('editor', 'blank_spaces')
             update_scrollbar = CONF.get('editor', 'scroll_past_end')
```
Let's make these options False in this PR.

goanpeca replied:
👍

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,549 @@
+
+
+# Constants
+LANGUAGE_EXTENSIONS = {
```
You're repeating here almost the same values present in the `EDIT_FILETYPES` dictionary, placed in `spyder/config/utils.py`. So please do a bit of refactoring in that file, to be able to use the same file extensions here.

goanpeca replied:
>So please do a bit of refactoring in that file, to be able to use the same file extensions here.

Well, it is not really the same and it would require doing some changes that really feel out of scope since currently the SimpleCodeEditor just depends on syntax highlighting and base. Now it will depend on some values on a utils file in a config module. 

This can be addressed in the Editor migration since all these values should probably live there anyway. I think these "config.utils" should be elsewhere so I prefer to address the changes in a holistic fashion he the work arrises.

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,549 @@
+    def __init__(self, code_editor=None):
+        super().__init__(code_editor)
+
+        self._editor = code_editor
+        self._left_padding = 6  # Pixels
+        self._right_padding = 3  # Pixels
```
`_left_padding` and `_right_padding` are used directly below to set them, so I don't think there's a need to make them start with and underscore `_`.

Also, is there a need to make `_editor` a private attribute?

goanpeca replied:
>so I don't think there's a need to make them start with and underscore

If it is not meant to be exposed to the user it should have an underscore.

>Also, is there a need to make _editor a private attribute?

Cleanliness. The line number area accesses the editor internally so there is no reason why this should be exposed.

ccordoba12 replied:
Ok

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,549 @@
+    # ------------------------------------------------------------------------
+    sig_focus_changed = Signal()
+    """
+    """
```
Missing docstring.

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,549 @@
+
+    def _highlight_current_line(self):
+        if self._highlighter:
+            extraSelections = []
```
Please don't use camel case completions. We don't use them anywhere else in our code.

goanpeca replied:
Fixed.

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,549 @@
+        if self._highlighter:
+            extraSelections = []
+            selection = QTextEdit.ExtraSelection()
+            lineColor = self._highlighter.get_currentline_color()
```
Also remove camel case here.

Besides, this simple editor needs to support cell highlighting so we can display that in the Appearance entry in Preferences.

goanpeca replied:
Fixed.

goanpeca commented:
>Looks good @goanpeca, thanks! I left some minor comments, but the most important thing is that this new SimpleCodeEditor needs to support cell highlighting so we can display it in Appearance. Also, perhaps in the future we could use cells in History too.

That is too much work for little gain. I do not agree.

ccordoba12 commented:
> That is too much work for little gain. I do not agree.

Cells is a very important functionality in Spyder (@isabela-pf remarked that in our last meeting with her). So we need to make it as visible as possible. Besides, as I told you, my plan is to use them in History too.

So please implement a simple version of them, or refactor and reuse what's necessary from `spyder/plugin/editor/widgets/base.py`.

goanpeca commented:
>Cells is a very important functionality in Spyder (@isabela-pf remarked that in our last meeting with her). So we need to make it as visible as possible. Besides, as I told you, my plan is to use them in History too.

I am not saying cells are not important. They are and there is a complex editor for that.

Using the complex editor in appearance is not an option as it results in a circular dependency.

Having a simple editor that is used in a lot of places for just syntax highlighting (I already agreed to add line numbers, cause I agree they are important) But cells are not for any of the current cases. Implementing cells (which is not important at the moment) is holding this PR and work that depends on it.

ccordoba12 commented:
Ok, let's go without cells for now, then we'll what to do about them later.

> If Spyder was configured to not start with the Editor on (which is something the new API allows for), what does it mean to have cells appear on the Theme selection?

I disagree on not having cells in Appearance, but let's not delay this work for that. Also, this argument makes no sense to me: what is Spyder if it only displays the theme selection plugin?

goanpeca commented:
>Also, this argument makes no sense to me: what is Spyder if it only displays the theme selection plugin?

A glorified Github issue reporter :-), where the appearance actually changed how the Issue Reporter will look :-)

goanpeca commented:
>Ok, let's go without cells for now, we'll see what to do about them later.

Thanks.

Please let's try to iterate faster on these PRs.

It is a big refactor, it is dealing with a lot of technical debt and is aiming to reduce the growth of the technical debt moving forward.

Yes, some things might break in the process (this is an alpha after all).

Yes, it is annoying that things break, but things need to break to move forward and I am ready to fix any issues that I may have introduced, but the faster things get merged, the faster we can iterate, the faster issues can be found and solved and the faster things will fall into place.

isabela-pf commented:
It looks like this has been resolved before I got a chance to comment, but I wanted to respond to being mentioned. I do think cells are important, but I think it is fine that a simple code editor has limited functionality, especially when it seems to exist mainly to solve some technical debt issues. Cells are still functioning elsewhere, so I think it's good to focus on solving the problems and working up to add things back in as they are a priority.

goanpeca commented:
Fixing issues.

goanpeca commented:
Ready for review @ccordoba12.

ccordoba12 commented:
/show binder

ccordoba12 commented on spyder/config/utils.py:
```diff
@@ -50,6 +50,28 @@
 ]
 
+
+LANGUAGE_EXTENSIONS = {
```
If you're not planning to use this dictionary in the SimpleCodeEditor, please remove it.

Else, please add a TODO comment above it to understand why you're putting it here and not using it.

goanpeca replied:
Forgot to remove it. Fixed.  (tracking this TODO on a separate issue already)

ccordoba12 commented on spyder/widgets/editor.py:
```diff
@@ -0,0 +1,551 @@
+# -*- coding: utf-8 -*-
```
Please rename this file to `simplecodeeditor.py`. `editor.py` is very vague and could lead to confusions later.

goanpeca replied:
Fixed.

ccordoba12 commented:
> Cells are still functioning elsewhere, so I think it's good to focus on solving the problems and working up to add things back in as they are a priority.

Thanks @isabela-pf for your feedback!

ccordoba12 commented:
This PR needs more work. There are still several issues that can be easily seen at plain sight (e.g. the font not being the same as the one used in the Editor).

@steff456 and @andfoy will manually review it and give you more concrete feedback.

goanpeca commented:
Thanks for the feedback @ccordoba12.

Looking forward to fixing any issues that @steff456 and @andfoy find!

steff456 commented:
Hi @goanpeca,
This are the comments we found in this PR

# History plugin
The first one is a screenshot of the editor and the second is a screenshot of the history. As you can see the font doesn't match.
![image](https://user-images.githubusercontent.com/20992645/90801572-b58f3d00-e2db-11ea-9927-13805ad0ff6c.png)

![image](https://user-images.githubusercontent.com/20992645/90801535-a4463080-e2db-11ea-82e8-c2cc3e7de6c8.png)

As well, we found that the debug bar is missing but we are not sure of this is intended or not, the first screenshot is from 4.x and the second one from this PR

![image](https://user-images.githubusercontent.com/20992645/90802582-1408eb00-e2dd-11ea-8bc5-1df1eef72a0f.png)

![image](https://user-images.githubusercontent.com/20992645/90802474-ee7be180-e2dc-11ea-9776-26941c045351.png)

# Help plugin
In here we noted that now the numbering of the editor is present, while in 4.x is not.
![image](https://user-images.githubusercontent.com/20992645/90803003-b6c16980-e2dd-11ea-9a6c-ba1607bf64eb.png)

![image](https://user-images.githubusercontent.com/20992645/90802966-ae692e80-e2dd-11ea-8846-885a83968be2.png)

Also, before we didn't have any highlight on the help and now we have. 

![image](https://user-images.githubusercontent.com/20992645/90804113-63501b00-e2df-11ea-91fd-51fc19c7aa1b.png)

![image](https://user-images.githubusercontent.com/20992645/90804153-706d0a00-e2df-11ea-9383-b1e23de43369.png)

# Appearance Dialog
At first sight we can see that you removed the `Run` config page, and I think that needs to be back. Also the order was changed.

![image](https://user-images.githubusercontent.com/20992645/90804909-80d1b480-e2e0-11ea-85d3-a5b1b23f13ca.png)

![image](https://user-images.githubusercontent.com/20992645/90804919-83cca500-e2e0-11ea-9a19-099cff15206b.png)

In the editor with the preview of the theme, we can see that there's no longer a cell which we believe is important to show that it has a different color than the rest. Also we found weird that even by writing a cell, the option is not activated. Additionally, the right vertical pane with the minimap is not shown. And, in this pane is happening exactly the same that in the history pane where the vertical pane of debug is not shown. All this things can be seen in the following screenshots.

![image](https://user-images.githubusercontent.com/20992645/90805294-0bb2af00-e2e1-11ea-9234-498fc7b27cd0.png)

![image](https://user-images.githubusercontent.com/20992645/90805356-1cfbbb80-e2e1-11ea-9cc2-d09b3f093754.png)

Also, we noted that by changing the theme the preview in editor is not shown, for example in this screenshot we should be seeing the preview of Pydev but we are still seeing Spyder Dark. Exactly the same behavior should be happening with the `edit selected scheme`.

![image](https://user-images.githubusercontent.com/20992645/90805564-664c0b00-e2e1-11ea-8d75-736cb9514fac.png)

# Error report
In this pane, in 4.x we had a vertical pane on the left side of the editor and now it is gone, we are not sure if that was intentional or not. As well, the highlight is present now, but it wasn't present before.

![image](https://user-images.githubusercontent.com/20992645/90806485-b37cac80-e2e2-11ea-9332-33a8c0b0c413.png)

![image](https://user-images.githubusercontent.com/20992645/90806508-bc6d7e00-e2e2-11ea-9359-b1d4e55667b9.png)

This were the things that we found on the appearance side that can be improved for this PR

goanpeca commented:
Thanks for the feedback @steff456 

# History plugin
>The first one is a screenshot of the editor and the second is a screenshot of the history. As you can see the font doesn't match.

Fixed.

>As well, we found that the debug panel is missing but we are not sure of this is intended or not, the first screenshot is from 4.x and the second one from this PR

That is expected. The SimpleCodeEditor only has line numbers, so not having a debug panel is normal. Also, there is no "debug"  functionality on the History plugin.

# Help plugin

>In here we noted that now the numbering of the editor is present, while in 4.x is not.

Fixed.

>Also, before we didn't have any highlight on the help and now we have. 

Fixed.

# Appearance Dialog

>At first sight we can see that you removed the `Run` config page, and I think that needs to be back. Also the order was changed.

That has been fixed already in master.

>In the editor with the preview of the theme, we can see that there's no longer a cell which we believe is important to show that it has a different color than the rest. Also we found weird that even by writing a cell, the option is not activated. 

This has been already addressed in the thread and will not be changed. The appearance plugin would need to depend on the editor plugin which creates a circular dependency. 

>Additionally, the right vertical pane with the minimap is not shown. And, in this pane is happening exactly the same that in the history pane where the vertical pane of debug is not shown. All this things can be seen in the following screenshots.

That is also expected as the Appearance configuration page is not using the CodeEditor from the Editor plugin, but a SimpleCodeEditor (what this or created) which only provides, syntax highlighting and the line numbers panel.

>Also, we noted that by changing the theme the preview in editor is not shown, for example in this screenshot we should be seeing the preview of Pydev but we are still seeing Spyder Dark. Exactly the same behavior should be happening with the `edit selected scheme`.

Fixed.

# Error report

>In this pane, in 4.x we had a vertical pane on the left side of the editor and now it is gone, we are not sure if that was intentional or not.

Yes, it is expected as this editor only provides a Line number area.

> As well, the highlight is present now, but it wasn't present before.

Fixed.

steff456 commented:
/show binder

goanpeca commented:
I already provided the fixes requested on this PR review.

dalthviz commented:
@andfoy @steff456 could you help us reviewing again this one to validate the fixes? Thanks!
